extensions/wikia/WikiaNewFiles/WikiaNewFiles.class.php

```diff
@@ -20,4 +65,147 @@ function execute( $par ) {
+		$cacheKey = wfMemcKey( __METHOD__, md5( $image->img_name ) );
+		$data = $wgMemc->get( $cacheKey );
+		if ( !is_array( $data ) ) {
+			// The ORDER BY ensures we get NS_MAIN pages first
+			$dbr = wfGetDB( DB_SLAVE );
+			$res = $dbr->select(
+				array( 'imagelinks', 'page' ),
+				array( 'page_namespace', 'page_title' ),
+				array( 'il_to' => $image->img_name, 'il_from = page_id' ),
+				__METHOD__,
+				array( 'LIMIT' => 2, 'ORDER BY' => 'page_namespace ASC' )
+			);
+
+			while ( $s = $res->fetchObject() ) {
+				$data[] = array( 'ns' => $s->page_namespace, 'title' => $s->page_title );
+			}
+			$dbr->freeResult( $res );
+
+			$wgMemc->set( $cacheKey, $data, 60 * 15 );
+		}
+
+		$links = array();
+
+		if ( !empty( $data ) ) {
+			foreach ( $data as $row ) {
+				$name = Title::makeTitle( $row['ns'], $row['title'] );
+				$links[] = Linker::link( $name, wfShortenText( $name, $anchorLength ), array( 'class' => 'wikia-gallery-item-posted' ) );
+			}
+		}
+
+		wfProfileOut( __METHOD__ );
+
+		return $links;
+	}
+
+	private function getPagination( $hidebots, $limit, $latestTimestamp, $hasOlderImages, $images ) {
+		global $wgLang, $wmu;
+
+		$lastImageIndex = count( $images ) - 1;
+		$firstTimestamp = wfTimestamp( TS_MW, $images[0]->img_timestamp );
+		$lastTimestamp = wfTimestamp( TS_MW, $images[$lastImageIndex]->img_timestamp );
+
+		$titleObj = SpecialPage::getTitleFor( 'Newimages' );
+
+		# If we change bot visibility, this needs to be carried along.
+		if ( !$hidebots ) {
+			$botpar = '&hidebots=0';
+		} else {
+			$botpar = '';
+		}
+		$now = wfTimestampNow();
+
+		/* @var $wgLang Language */
+		$d = $wgLang->date( $now, true );
+		$t = $wgLang->time( $now, true );
+		$dateLink = Linker::link( $titleObj, wfMessage( 'sp-newimages-showfrom' )->rawParams( $d, $t )->escaped(),
+			array( 'class' => 'navigation-filesfrom' ),
+			'from=' . $now . $botpar );
+
+		$botLink = Linker::link( $titleObj,
+			wfMessage( 'showhidebots', $hidebots ? wfMessage( 'show' )->plain() : wfMessage( 'hide' )->plain() )->escaped(),
+			array( 'class' => 'navigation-' . ( $hidebots ? 'showbots' : 'hidebots' ) ),
+			'hidebots=' . ( $hidebots ? '0' : '1' ) );
+
+		$prevLink = wfMessage( 'pager-newer-n', $wgLang->formatNum( $limit ) )->escaped();
+		if ( $firstTimestamp && $firstTimestamp != $latestTimestamp ) {
+			$wmu['prev'] = $firstTimestamp;
+			$prevLink = Linker::link( $titleObj, $prevLink,
+				array( 'class' => 'navigation-newer' ),
+				'from=' . $firstTimestamp . $botpar );
+		}
+
+		$nextLink = wfMessage( 'pager-older-n', $wgLang->formatNum( $limit ) )->escaped();
+		if ( $hasOlderImages && $lastTimestamp ) {
+			$wmu['next'] = $lastTimestamp;
+			$nextLink = Linker::link( $titleObj, $nextLink,
+				array( 'class' => 'navigation-older' ),
+				'until=' . $lastTimestamp . $botpar );
+		}
+
+		return '<p id="newfiles-nav">' . $botLink . ' ' . wfMessage( 'viewprevnext' )->rawParams( $prevLink, $nextLink, $dateLink )->escaped() . '</p>';
```

Comment:
Might be nice to format this line (and a few others) to multiple lines so it doesn't scroll off the side of the screen.